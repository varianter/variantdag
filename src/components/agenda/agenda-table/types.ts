import { FeatureSlot } from './feature-slot/FeatureSlot';
import { RowSlot } from './row-slot/RowSlot';

/**
 * Duration created from hours and minutes used for relative time calculation.
 *
 * Example: 10:30 => 10*60 minutes + 30 minutes = 630 minutes
 */
export type DurationMinutes = number;

export type Time = {
  hour: number;
  minutes: number;
};

/**
 * Location of a feature, such as a presentation or meeting:
 *
 * Examples: "Oktetten", "Amfiet"
 */
export type VenueName = string;

/**
 * A time slot in the program
 */
export interface Slot {
  from: Time;
  to: Time;
}

/**
 * The entire program of a conference. The variables 'from' and 'to' describes
 * the start and end times of the program, and is necessary for style calculations.
 *
 * The base interfaces FeatureSlot and RowSlot contains the properties required by AgendaTable.
 * However, in order to allow writing render functions for Row and Feature slots, the base
 * interfaces may be extended and used instead.
 */
export type Program<
  T extends FeatureSlot = FeatureSlot,
  K extends RowSlot = RowSlot,
> = {
  from: Time;
  to: Time;
  features: T[];
  rows: K[];
};

/**
 * Any function that implements the render functions, can be used to control how
 * rows, features and headers are displayed in the AgendaTable. Extend the RowSlot and FeatureSlot
 * interfaces to allow rendering with additional properties.
 */
export type RowRenderer<T extends RowSlot> = (row: T) => JSX.Element;

export type FeatureRenderer<T extends FeatureSlot> = (
  feature: T,
) => JSX.Element;

export type HeaderRenderer = (venueName: VenueName) => JSX.Element;
