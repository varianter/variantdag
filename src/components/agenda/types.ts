/**
 * Duration created from hours and minutes used for relative time calculation.
 *
 * Example: 10:30 => 10*60 minutes + 30 minutes = 630 minutes
 */
export type DurationMinutes = number;

export type Time = {
  hour: number;
  minutes: number;
};

/**
 * Location of a feature, such as a presentation or meeting:
 *
 * Examples: "Oktetten", "Amfiet"
 */
export type Venue = string; // e.g

/**
 * A time slot in the program
 */
export interface Slot {
  from: Time;
  to: Time;
}

/**
 * A presentation, meeting or similar
 */
export interface Feature extends Slot {
  title: string;
  venue: Venue;
}

/**
 * The entire program of a conference. The variables 'from' and 'to' describes
 * the start and end times of the program, and is necessary for style calculations.
 */
export type Program = {
  from: Time;
  to: Time;
  features: Feature[];
};
