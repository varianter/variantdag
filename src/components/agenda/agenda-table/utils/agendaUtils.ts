import { DurationMinutes, Feature, Program, Slot, Time } from '../types';

export const getUniqueVenues = (features: Feature[]) =>
  features
    .map((feature) => feature.venue)
    .filter((venue, index, venues) => venues.indexOf(venue) === index);

export const asDuration = (time: Time) => time.hour * 60 + time.minutes;

export const percentageOf = (
  duration: DurationMinutes,
  totalDuration: DurationMinutes,
) => (duration / totalDuration) * 100;

export const getDuration = (program: Program): DurationMinutes =>
  asDuration(program.to) - asDuration(program.from);

export const calculateSlotPositionStyle = (program: Program, slot: Slot) => {
  const programDuration = getDuration(program);
  const slotDuration = asDuration(slot.to) - asDuration(slot.from);

  return {
    height: `${(slotDuration / programDuration) * 100}%`,
    top: `${
      ((asDuration(slot.from) - asDuration(program.from)) / programDuration) *
      100
    }%`,
  };
};
