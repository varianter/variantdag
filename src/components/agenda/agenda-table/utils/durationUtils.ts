import { DurationMinutes, Program, Slot, Time } from '../types';

export const asDuration = (time: Time) => time.hour * 60 + time.minutes;

export const percentageOf = (
  duration: DurationMinutes,
  totalDuration: DurationMinutes,
) => (duration / totalDuration) * 100;

export const getDuration = (program: Program): DurationMinutes =>
  asDuration(program.to) - asDuration(program.from);

export const calculateSlotPositionStyle = (program: Program, slot: Slot) => {
  const programDuration = getDuration(program);
  const slotDuration = asDuration(slot.to) - asDuration(slot.from);

  return {
    height: `${(slotDuration / programDuration) * 100}%`,
    top: `${
      ((asDuration(slot.from) - asDuration(program.from)) / programDuration) *
      100
    }%`,
  };
};
