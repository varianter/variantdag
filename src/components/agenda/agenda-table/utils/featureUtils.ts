import { FeatureSlot } from '../feature-slot/FeatureSlot';

export const getUniqueVenues = (features: FeatureSlot[]) =>
  features
    .map((feature) => feature.venueName)
    .filter((venue, index, venues) => venues.indexOf(venue) === index);
