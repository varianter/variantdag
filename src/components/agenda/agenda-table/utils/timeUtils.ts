import { Time } from '../types';

export const time = (time: string): Time => {
  const [hour, minutes] = time.split(':').map(Number);
  return {
    hour: hour,
    minutes: minutes,
  };
};

export const format = (time: Time): string => {
  const hour = time.hour < 10 ? `0${time.hour}` : time.hour;
  const minutes = time.minutes < 10 ? `0${time.minutes}` : time.minutes;

  return `${hour}:${minutes}`;
};
