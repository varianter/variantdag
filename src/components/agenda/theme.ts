import { colorPairs } from '@variant/profile/lib/colors';
import { CSSProperties } from 'react';

export type Color = string;

/**
 * Describes how a Row or Feature should look
 */
export interface ThemeType {
  text: Color;
  background: Color;
}

export const Theme = {
  FELLES: {
    text: colorPairs.primary.default.text,
    background: colorPairs.primary.default.bg,
  } as ThemeType,
  LÃ†RE: {
    text: colorPairs.secondary1.default.text,
    background: colorPairs.secondary1.default.bg,
  } as ThemeType,
  SKUDD: {
    text: colorPairs.secondary2.default.text,
    background: colorPairs.secondary2.default.bg,
  } as ThemeType,
  ANNET: {
    text: colorPairs.secondary4.default.text,
    background: colorPairs.secondary4.default.bg,
  } as ThemeType,
  PAUSE: {
    text: colorPairs.secondary3.default.text,
    background: colorPairs.secondary3.default.bg,
  } as ThemeType,
};

export type ThemeName = keyof typeof Theme;

export const toStyle = (theme: ThemeType): CSSProperties => ({
  color: theme.text,
  backgroundColor: theme.background,
});
