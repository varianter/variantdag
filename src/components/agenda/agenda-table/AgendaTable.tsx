import { FC } from 'react';
import styles from './AgendaTable.module.css';
import { Venue } from './venue/Venue';
import { FeatureSlot } from './feature-slot/FeatureSlot';
import {
  FeatureRenderer,
  HeaderRenderer,
  Program,
  RowRenderer,
  VenueName,
} from './types';
import { calculateSlotPositionStyle } from './utils/durationUtils';
import { getUniqueVenues } from './utils/featureUtils';
import { RowSlot } from './row-slot/RowSlot';
import { VenueHeader } from './venue/VenueHeader';
import { format } from './utils/timeUtils';

const generateVenueHeaders = (
  uniqueVenues: VenueName[],
  renderVenueHeader: HeaderRenderer,
) =>
  uniqueVenues.map((venue) => (
    <VenueHeader key={venue}>{renderVenueHeader(venue)}</VenueHeader>
  ));

const generateVenues = <T extends FeatureSlot, K extends RowSlot>(
  uniqueVenues: VenueName[],
  program: Program<T, K>,
  renderFeature: FeatureRenderer<T>,
) =>
  uniqueVenues.map((venueName) => {
    const featuresInVenue = program.features.filter(
      (feature) => feature.venueName === venueName,
    );

    const features = featuresInVenue.map((feature) => (
      <FeatureSlot
        key={`${format(feature.from)}-${format(feature.to)}-${
          feature.venueName
        }`}
        style={calculateSlotPositionStyle(program, feature)}
      >
        {renderFeature(feature)}
      </FeatureSlot>
    ));

    return <Venue key={venueName}>{features}</Venue>;
  });

const generateRows = <T extends FeatureSlot, K extends RowSlot>(
  program: Program<T, K>,
  renderRow: RowRenderer<K>,
) =>
  program.rows.map((row: K) => (
    <RowSlot
      key={`${format(row.from)}-${format(row.to)}`}
      style={calculateSlotPositionStyle(program, row)}
      rowSlot={row}
    >
      {renderRow(row)}
    </RowSlot>
  ));

type Props<T extends FeatureSlot, K extends RowSlot> = {
  program: Program<T, K>;
  height: string;
  renderFeature: FeatureRenderer<T>;
  renderRow: RowRenderer<K>;
  renderVenueHeader: HeaderRenderer;
};

/**
 * Creates a Agenda with Venues (e.g 'Amfiet') containing one or more Features (e.g talks, presentations or workshops).
 * An Agenda may optionally include Rows - time slots that span accross all Venues - such as breaks or common events.
 * All Features and Rows are positioned relative to the overall height of the Agenda and may overlap.
 *
 * @param program - contains metadata for features and rows to be displayed in the agenda
 * @param height - the overall height of the agenda (e.g '1000px'). All agenda features will scale relative to this height.
 * @param renderFeature - determines how each feature is rendered
 * @param renderRow - determeines how each row is rendered
 * @param renderVenueHeader - determines how each venue header is rendered
 */
export const AgendaTable = <T extends FeatureSlot, K extends RowSlot>({
  program,
  height,
  renderFeature,
  renderRow,
  renderVenueHeader,
}: Props<T, K>) => {
  const uniqueVenues = getUniqueVenues(program.features);
  const venueHeaders = generateVenueHeaders(uniqueVenues, renderVenueHeader);
  const venues = generateVenues(uniqueVenues, program, renderFeature);
  const rowSlots = generateRows(program, renderRow);

  return (
    <article style={{ height: height }} className={styles.agenda}>
      <header className={styles.venueHeaders}>{venueHeaders}</header>
      <main className={styles.content}>
        <div className={styles.rowSlots}>{rowSlots}</div>
        <div className={styles.venues}>{venues}</div>
      </main>
    </article>
  );
};
