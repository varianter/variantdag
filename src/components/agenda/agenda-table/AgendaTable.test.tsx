import renderer from 'react-test-renderer';
import { Theme } from '../theme';
import { AgendaTable } from './AgendaTable';
import { Program } from './types';
import { time } from './utils/timeUtils';

export const simpleProgram: Program = {
  from: time('12:30'),
  to: time('16:00'),
  features: [
    {
      title: 'Samarbeidscage for design',
      venue: 'Digs 5. egt - East',
      from: time('12:30'),
      to: time('13:45'),
      theme: Theme.FELLES,
    },
    {
      title: 'Planlegging av skudd: sky',
      venue: 'Digs 5. etg - West',
      from: time('13:00'),
      to: time('14:45'),
      theme: Theme.PAUSE,
    },
    {
      title: 'Bærekraft og OKR',
      venue: 'Lounge på Digs',
      from: time('12:30'),
      to: time('14:45'),
      theme: Theme.LÆRE,
    },
    {
      title: 'OKR Q4 \n Selskapsstatus',
      venue: 'Digs 5. egt - East',
      from: time('15:00'),
      to: time('16:00'),
      theme: Theme.SKUDD,
    },
  ],
  rows: [
    {
      from: time('14:45'),
      to: time('15:00'),
      description: 'Pause',
      theme: Theme.ANNET,
    },
  ],
};

const renderBox = (content: any) => <div>{JSON.stringify(content)}</div>;

describe('AgendaTable', () => {
  it('Renders slots of correct size', () => {
    const agenda = renderer.create(
      <AgendaTable
        program={simpleProgram}
        renderFeature={renderBox}
        renderRow={renderBox}
        renderVenueHeader={renderBox}
        height={'1200px'}
      />,
    );

    expect(agenda).toMatchSnapshot();
  });
});
