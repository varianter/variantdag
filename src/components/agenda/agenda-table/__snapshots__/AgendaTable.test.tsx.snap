// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgendaTable Renders slots of correct size 1`] = `
<article
  className="agenda"
  style={
    Object {
      "height": "1200px",
    }
  }
>
  <header
    className="venueHeaders"
  >
    <div
      className="header"
    >
      <div>
        "Digs 5. egt - East"
      </div>
    </div>
    <div
      className="header"
    >
      <div>
        "Digs 5. etg - West"
      </div>
    </div>
    <div
      className="header"
    >
      <div>
        "Lounge på Digs"
      </div>
    </div>
  </header>
  <main
    className="content"
  >
    <div
      className="rowSlots"
    >
      <div
        className="rowSlot"
        style={
          Object {
            "height": "7.142857142857142%",
            "top": "64.28571428571429%",
          }
        }
      >
        <div>
          {"from":{"hour":14,"minutes":45},"to":{"hour":15,"minutes":0}}
        </div>
      </div>
    </div>
    <div
      className="venues"
    >
      <div
        className="venue"
      >
        <div
          className="featureSlots"
        >
          <div
            className="featureSlot"
            style={
              Object {
                "height": "35.714285714285715%",
                "top": "0%",
              }
            }
          >
            <div>
              {"venueName":"Digs 5. egt - East","from":{"hour":12,"minutes":30},"to":{"hour":13,"minutes":45}}
            </div>
          </div>
          <div
            className="featureSlot"
            style={
              Object {
                "height": "28.57142857142857%",
                "top": "71.42857142857143%",
              }
            }
          >
            <div>
              {"venueName":"Digs 5. egt - East","from":{"hour":15,"minutes":0},"to":{"hour":16,"minutes":0}}
            </div>
          </div>
        </div>
      </div>
      <div
        className="venue"
      >
        <div
          className="featureSlots"
        >
          <div
            className="featureSlot"
            style={
              Object {
                "height": "50%",
                "top": "14.285714285714285%",
              }
            }
          >
            <div>
              {"venueName":"Digs 5. etg - West","from":{"hour":13,"minutes":0},"to":{"hour":14,"minutes":45}}
            </div>
          </div>
        </div>
      </div>
      <div
        className="venue"
      >
        <div
          className="featureSlots"
        >
          <div
            className="featureSlot"
            style={
              Object {
                "height": "64.28571428571429%",
                "top": "0%",
              }
            }
          >
            <div>
              {"venueName":"Lounge på Digs","from":{"hour":12,"minutes":30},"to":{"hour":14,"minutes":45}}
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</article>
`;
